cpp/src/copying/scatter.cu

```diff
@@ -0,0 +1,381 @@
+  std::unique_ptr<column> operator()(column_view const& source,
+      MapIterator scatter_iter, size_type scatter_rows, column_view const& target,
+      rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+  {
+    auto result = std::make_unique<column>(target, stream, mr);
+    auto result_view = result->mutable_view();
+
+    // NOTE use source.begin + scatter rows rather than source.end in case the
+    // scatter map is smaller than the number of source rows
+    thrust::scatter(rmm::exec_policy(stream)->on(stream), source.begin<T>(),
+      source.begin<T>() + scatter_rows, scatter_iter,
+      result_view.begin<T>());
+
+    return result;
+  }
+
+  template <typename T, std::enable_if_t<not is_fixed_width<T>()>* = nullptr>
+  std::unique_ptr<column> operator()(column_view const& source,
+      MapIterator scatter_iter, size_type scatter_rows, column_view const& target,
+      rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+  {
+    using strings::detail::create_string_vector_from_column;
+    auto const source_vector = create_string_vector_from_column(source, stream);
+    auto const begin = source_vector.begin();
+    auto const end = begin + scatter_rows;
+    return strings::detail::scatter(begin, end, scatter_iter, target, mr, stream);
+  }
+};
+
+struct scatter_impl {
+  template <typename T, std::enable_if_t<std::is_integral<T>::value
+     and not std::is_same<T, bool8>::value>* = nullptr>
+  std::unique_ptr<table> operator()(
+      table_view const& source, column_view const& scatter_map,
+      table_view const& target, bool check_bounds,
+      rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+  {
+    if (check_bounds) {
+      auto const begin = -target.num_rows();
+      auto const end = target.num_rows();
+      auto bounds = bounds_checker<T>{begin, end};
+      CUDF_EXPECTS(thrust::all_of(rmm::exec_policy(stream)->on(stream),
+        scatter_map.begin<T>(), scatter_map.end<T>(), bounds),
+        "Scatter map index out of bounds");
+    }
+
+    // Transform negative indices to index + target size
+    auto scatter_rows = scatter_map.size();
+    auto scatter_iter = thrust::make_transform_iterator(
+      scatter_map.begin<T>(), index_converter<T>{target.num_rows()});
+
+    // Second dispatch over data type per column
+    auto result = std::vector<std::unique_ptr<column>>(target.num_columns());
+    auto scatter_functor = column_scatterer<decltype(scatter_iter)>{};
+    std::transform(source.begin(), source.end(), target.begin(), result.begin(),
+      [=](auto const& source_col, auto const& target_col) {
+        return type_dispatcher(source_col.type(), scatter_functor,
+          source_col, scatter_iter, scatter_rows, target_col, mr, stream);
+      });
+
+    auto gather_map = make_gather_map<T>(scatter_iter, scatter_rows,
+      target.num_rows(), stream);
+    gather_bitmask(source, gather_map.begin(), result, mr, stream);
+
+    return std::make_unique<table>(std::move(result));
```

Comment:
Is the `std::move` necessary here?  I think I remember @jrhemstad saying this prevents return value optimization?

Reply:
I think it's required here since this particular table constructor takes `table(std::vector<std::unique_ptr<column>>&& columns)`